core/src/main/java/org/apache/calcite/plan/volcano/VolcanoRuleCall.java

```diff
@@ -190,7 +190,8 @@ protected void onMatch() {
           return;
         }
 
-        if (subset.set.equivalentSet != null) {
+        if ((subset.set.equivalentSet != null)
+            || (subset != rel && !subset.getRelList().contains(rel))) {
```

Comment:
Could you please explain in what scenarios does the following condition hold?
`!subset.getRelList().contains(rel)`

Reply:
when we merge set, the rel may be obsolete, but the rule match is still in the rule queue. Later I'll try to contrast a ut to cover this.

Reply:
Thanks for your feedback. When merging set, the condition should already been covered by `subset.set.equivalentSet != null`?

Reply:
Latter I'll debug again, we found this in our product env.

Reply:
@liyafan82 I dig the code, and it's not about equivalentSet, let me explain to you:
1. In RelSet#mergeWith, we ill rename parent rels. note, what we renamed is the equivalentSet's parent, so it may not have equivalentSet(child have equivalentSet)

```
    // Update all rels which have a child in the other set, to reflect the
    // fact that the child has been renamed.
    //
    // Copy array to prevent ConcurrentModificationException.
    final List<RelNode> previousParents =
        ImmutableList.copyOf(otherSet.getParentRels());
    for (RelNode parentRel : previousParents) {
      planner.rename(parentRel);
    }
```

2.

Reply:
@Aaaaaaron Thanks for the clarification. It would be nice if we could add a brief description in the comment.

Reply:
> @Aaaaaaron Thanks for the clarification. It would be nice if we could add a brief description in the comment.

Updated, could you please take a look?

Reply:
Looks good. Thanks.
This PR seems like a positive change.

Reply:
The operation `subset.getRelList().contains(rel)` is with poor performance. Can you give evidence how much gains of the patch ?

Reply:
Hi @danny0405 , sorry for the late reply. We found this case in our product env and it been so long that I can not find the origin case 😂, I'll try again.
By the performance, we had monitor, it does not increase the planning time.

Reply:
For the potential performance issue, how about I add a method `getRelSet`, the set#contains's performance should be better than list#contain

Reply:
Thanks for the explanation, i'm fine with the change.

Reply:
This patch results in 40% CPU increase in `org.apache.calcite.adapter.tpch.TpchTest#testQuery07` :(